Fixed avro binary serialization/deserialization

annovi opened the pull request:
issue #21 
- removed the flag "include schema" from the AvroOutputStream because according with the Apache Avro documentation the schema is always included with the Data. So it's not need it.
[http://avro.apache.org/docs/1.8.0/gettingstartedjava.html: "Data in Avro is always stored with its corresponding schema, meaning we can always read a serialized item regardless of whether we know the schema ahead of time."
- added "AvroBinaryInputStream" and "AvroDataInputStream" that extends AvroInputStream, in order to distinguish between data file stream or binary stream
- fixed the Unit tests

sksamuel commented:
Thanks. I have a couple of comments.

1. Can you get all tests to pass after this PR?
2. It's allowed to not include the schema in the data. Although this is not the _normal_ case, it is supported, so I don't think we should remove the `includeSchema` flag. Maybe the flag should be renamed to `binaryMode`, or maybe people can create the different streams programatically (ie, no super trait). See https://thetechsolo.wordpress.com/2015/01/17/apache-avro-schema-less-serialization-how-to/comment-page-1/
3. Can you add a unit test for reading multiple values from a binary file, using your new Binary input stream, because I was unable to get it to work.

Thanks again.

annovi commented:
1) The only test that failed was the "UniverseTest.scala". Sorry,  I am going to fix it.
2) The problem with the previous solution is that when you want to serialize using **AvroBinaryOutputStream** you have to set the flag `includeSchema` to `false`. 
Let's take a look at the **AvoutOutputStream.scala** unitest:
`
   line 15:  val output = AvroOutputStream[Composer](baos, false)
`
In this example said: I want to serialize without schema (flag includeSchema = false), but then I pass the schema anyway (Composer is the schema). So I found that a bit strange.
As you suggested we can set the flag `binaryMode` to `true`, in case you want to use binary serialization. The problem with this approach (might be that I'm wrong) when you deserialize calling **AvroInputStream** we should known in advance whether to use the flag `binaryMode` to `true` or not, while most of the times you got a **serialized avro message** from an application, but you are not aware of the fact that is binary or not. If we want to deserialize a message that before was serialized as binary, we should called with a second parameter set to `true`. Anyway I open to your suggestion,  and possibile alternative solutions/implementation.
3) We do not need an additional unit test for the the version of the **AvroInputStream** because the current unit test **AvroInputStreamTest** is using feeding the class AvroInputStream with `array[Byte]` therfore if is already using the new version of the AvroBinaryInputStream that I made. The test in the  **AvroInputStreamTest** are already serializing/deserializing multiple values from a binary file.

sksamuel commented:
Hi,

1. Ok
2. As long as we support binary in output streams, I don't mind how its achieved. Although if we just rename the parameter to binaryMode, then it means existing code will still compile but be reversed, so perhaps that won't work. Perhaps just making both the classes directly creatable is the answer, eg
new AvroDataOutputStream(file) or whatever.

I have no problem with the fact its up to the user to know if their incoming stream has the schema embedded or not. That is something they would decide in their protocol

3. I mean a unit test that uses an input that has 2 or more case classes in a binary stream, not this:

    "read byte arrays" in {
      val data = Seq(ByteArrayTest(Array[Byte](1, 2, 3)), ByteArrayTest(Array[Byte](125, 126, 127)))
      val bytes = write(data)

      val in = AvroInputStream[ByteArrayTest](bytes)
      val result = in.iterator.toList
      result.head.bytes.toList shouldBe Array[Byte](1, 2, 3).toList
      result.last.bytes.toList shouldBe Array[Byte](125, 126, 127).toList
      in.close()
    }

Which is only testing that we can serialize byte arrays.

annovi commented:
- Fixed the UniverseTest.scala that was broken.
- Removed the AvroInputDataStream because is not need it.
- Restored the AvroOutputStream as it was before, with the possibility to use an additional parameter to enable/disable binary mode serialization. 

TODO: add an other Unit Test as you mention in the point 3)

annovi commented:
I update the pull request.

annovi commented:
Did you take a look at my latest PR? What do you think?

sksamuel commented:
I'll check it out this weekend

On 18 May 2016 at 08:35, Andrea Annovi <notifications@github.com> wrote:

> Did you take a look at my latest PR? What do you think?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/sksamuel/avro4s/pull/23#issuecomment-219948921>
>

sksamuel commented:
It looks good. Only change I think is to make it so that it doesn't load all the records at the start, otherwise having an iterator is pointless and it might as well just return a Seq. I can make that change once merged in though.

annovi commented:
Thank you. That is great!
